sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs

```diff
@@ -1667,6 +1665,156 @@ public BlobServiceClient(Uri serviceUri, TokenCredential credential, BlobClientO
+        /// </remarks>
+        public virtual async Task<Response<BlobContainerClient>> RenameBlobContainerAsync(
+            string destinationContainerName,
+            string sourceContainerName,
+            BlobRequestConditions sourceConditions = default,
+            CancellationToken cancellationToken = default)
+            => await RenameBlobContainerInternal(
+                destinationContainerName,
+                sourceContainerName,
+                sourceConditions,
+                async: true,
+                cancellationToken)
+                .ConfigureAwait(false);
+
+        /// <summary>
+        /// Renames an existing Blob Container.
+        /// </summary>
+        /// <param name="destinationContainerName">
+        /// The new name of the Blob Container.
+        /// </param>
+        /// <param name="sourceContainerName">
+        /// The name of the source container.
+        /// </param>
+        /// <param name="sourceConditions">
+        /// Optional <see cref="BlobRequestConditions"/> to add
+        /// conditions on the renaming of this container.
+        /// </param>
+        /// <param name="async">
+        /// Whether to invoke the operation asynchronously.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{BlobContainerClient}"/> pointed at the renamed container.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        internal async Task<Response<BlobContainerClient>> RenameBlobContainerInternal(
+            string destinationContainerName,
+            string sourceContainerName,
+            BlobRequestConditions sourceConditions,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            using (Pipeline.BeginLoggingScope(nameof(BlobContainerClient)))
+            {
+                Pipeline.LogMethodEnter(
+                    nameof(BlobContainerClient),
+                    message:
+                    $"{nameof(Uri)}: {Uri}\n" +
+                    $"{nameof(sourceContainerName)}: {sourceContainerName}");
+
+                try
+                {
+                    BlobContainerClient containerClient = GetBlobContainerClient(destinationContainerName);
+
+                    Response response = await BlobRestClient.Container.RenameAsync(
+                        clientDiagnostics: ClientDiagnostics,
+                        pipeline: Pipeline,
+                        resourceUri: containerClient.Uri,
+                        version: Version.ToVersionString(),
+                        sourceContainerName: sourceContainerName,
+                        sourceLeaseId: sourceConditions?.LeaseId,
```

Comment:
so we only take lease from source conditions? in that case we should mention that in docs.

Reply:
The [spec](https://microsoft.sharepoint.com/:w:/t/AzureStorage/EX0crOv7mDROnz68USoD4xoBV71HDPplg1RKbU9l7jv8PQ?e=FnwILy&CID=C0E3F63A-6196-4B76-A1C9-5EFBBF641189&wdLOR=c40ED4F1D-18CB-4C31-8393-8BF49A8EC12F) only mentions `x-ms-source-lease-id`.  I think this is self documenting, since `destinationAccessConditions` is not a parameter.

Reply:
What I meant is that we only honor `LeaseId` from `BlobRequestConditions`. Other properties are ineffective in this scenario. We should mention that.

Reply:
Fixed.